src/main/java/org/duke/Duke.java

```diff
@@ -0,0 +1,199 @@
+package org.duke;
+
+import org.duke.cmd.Command;
+import org.duke.cmd.CommandDispatcher;
+import org.duke.json.JsonParser;
+import org.duke.json.JsonWriter;
+import org.duke.json.ValueHandler;
+import org.duke.task.DeadlineTask;
+import org.duke.task.EventTask;
+import org.duke.task.Task;
+import org.duke.task.TaskType;
+import org.duke.ui.DukeIO;
+import org.duke.util.CounterDecorator;
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+
+/**
+ * Main class for Duke.
+ */
+public class Duke {
+
+    private static final String[] initialGreeting = new String[]{
+            "Hello! I'm Duke",
+            "What can I do for you?"
+    };
+    private static final String SAVE_PATH = "./duke.json";
+    private final CommandDispatcher dispatcher;
+    private final DukeIO io;
+    private ArrayList<Task> taskList;
+
+    public Duke(DukeIO io) {
+        this.io = io;
+
+        this.dispatcher = new CommandDispatcher();
+        //Bind command handlers
+        this.dispatcher.bindCommand("list", this::displayList);
+        this.dispatcher.bindCommand("done", this::markAsDone);
+        this.dispatcher.bindCommand("bye", this::handleBye);
+        this.dispatcher.bindCommand("deadline", this::makeDeadlineTask);
+        this.dispatcher.bindCommand("event", this::makeEventTask);
+        this.dispatcher.bindCommand("todo", this::makeToDoTask);
+        this.dispatcher.bindCommand("delete", this::deleteTask);
+        this.dispatcher.bindCommand("find", this::findTasks);
+        this.dispatcher.setUnknownCommandHandler(cmd -> {
+            throw new DukeException("I'm sorry, but I don't know what that means. :-(");
+        });
+        this.io.setCommandDispatcher(this.dispatcher);
+
```

Comment:
I like how you implemented all your commands with a central dispatcher. It is very neat and shows clearly what text is bound to each command. Also like how you had a default fallback command handler to handle unknown commands from user input.